### PR TITLE
[Impeller] Write a text-decoration test at the `dart:ui` layer

### PR DESCRIPTION
**Pull request by matanlurey**
Closes https://github.com/flutter/flutter/issues/138501.

---

It would be nice to turn this into a test-fixture, and create a better local environment.

Here is how to run it locally:

```bash
# The test expects you to be PWD in the `src` directory or it crashes.
cd $ENGINE_SRC

out/host_debug_unopt_arm64/flutter_tester \
  --force-multithreading \
  --disable-observatory \
  --use-test-fonts \
  --icu-data-file-path=out/host_debug_unopt_arm64/icudtl.dat \
  --flutter-assets-dir=out/host_debug_unopt_arm64/gen/flutter/lib/ui/assets \
  --disable-asset-fonts \
  --enable-impeller \
  out/host_debug_unopt_arm64/gen/canvas_test.dart.dill

# "See" the output goldens.
open out/host_debug_unopt_arm64/gen/skia_gold_canvas_test.dart
```

image

### CODE REVIEW

**Comment by jonahwilliams** [comment]
It looks like the text isn't showing up in the goldens ...

File: testing/dart/canvas_test.dart [hunk 1]
```diff
@@ -1031,6 +1031,38 @@ void main() async {
     canvas.restoreToCount(canvas.getSaveCount() + 1);
     expect(canvas.getSaveCount(), equals(6));
   });
+
+  test('TextDecoration renders non-solid lines', () async {
```
Comment (jonahwilliams) [inline testing/dart/canvas_test.dart:1035]: This doesn't seem that bad, you made it seem like this would be pretty hard

Comment (matanlurey) [reply testing/dart/canvas_test.dart:1035]: Easy to write, hard to run :)

File: testing/dart/canvas_test.dart [hunk 2]
```diff
@@ -1031,6 +1031,38 @@ void main() async {
...
+    final File file = File(path.join('flutter', 'testing', 'resources', 'RobotoSlab-VariableFont_wght.ttf'));
```
Comment (jonahwilliams) [inline testing/dart/canvas_test.dart:1036]: So I bet this path is wrong. There is some logic in here to look up the out directory from a system environment variable. We use that in some of the shader tests, let me find it

**Comment by matanlurey** [comment]
> It looks like the text isn't showing up in the goldens ...

Yeah I'm guessing this is a PWD issue similar to what I was seeing locally - test assets are expected to be in a specific place. Let's look at this together tomorrow, unless @dnfield notices I did something wrong.

File: testing/dart/canvas_test.dart (hunk 2 above)
Comment (jonahwilliams) [reply testing/dart/canvas_test.dart:1036]: See https://github.com/flutter/engine/blob/main/testing/dart/shader_test_file_utils.dart

You can get the build directory from the system env variable, and then the fixtures should be available under that.

Comment (matanlurey) [reply testing/dart/canvas_test.dart:1036]: Well, that makes the local build even more annoying (another thing I have to magically set or it fails). But thanks!

File: testing/dart/canvas_test.dart [hunk 3]
```diff
@@ -124,6 +124,14 @@ void testNoCrashes() {
   });
 }
 
+String get _flutterBuildPath {
+  final String? buildPath = Platform.environment['FLUTTER_BUILD_DIRECTORY'];
+  if (buildPath == null) {
+    throw StateError('FLUTTER_BUILD_DIRECTORY environment variable is not set.');
+  }
+  return buildPath;
+}
```
Comment (dnfield) [inline testing/dart/canvas_test.dart:133]: This is probably fine, but you could alternatively write this as an environment define like how I did for the Skia gold output directory, see

https://github.com/flutter/engine/blob/ed5cf7f5427c7706873a58a5ed972f0a0da7af43/testing/dart/BUILD.gn#L55

Comment (dnfield) [reply testing/dart/canvas_test.dart:133]: (that way there's no extra environment variables to set, although it does still require some magic in the compilation)

**Comment by matanlurey** [comment]
New code up, should work both locally and on CI :)

File: testing/dart/BUILD.gn [hunk 4]
```diff
@@ -52,8 +52,13 @@ foreach(test, tests) {
   flutter_frontend_server("compile_$test") {
     main_dart = test
     kernel_output = "$root_gen_dir/$test.dill"
-    extra_args = [ "-DkSkiaGoldWorkDirectory=$skia_gold_work_dir" ]
+    extra_args = [
+      "-DkFlutterBuildDirectory=$root_gen_dir",
```
Comment (dnfield) [inline testing/dart/BUILD.gn:56]: Use `rebase_path` on this like with the skia_gold_work_dir above so you get a non-gn-ified path.

File: testing/dart/canvas_test.dart [hunk 5]
```diff
@@ -124,6 +124,19 @@ void testNoCrashes() {
...
+const String kFlutterBuildDirectory = 'kFlutterBuildDirectory';
+
+String get _flutterBuildPath {
+  const String buildPath = String.fromEnvironment(kFlutterBuildDirectory);
+  if (buildPath.isEmpty) {
+    throw StateError('kFlutterBuildDirectory -D variable is not set.');
+  }
+  if (buildPath.startsWith('//')) {
+    return buildPath.substring(2);
+  }
```
Comment (dnfield) [inline testing/dart/canvas_test.dart:136]: If you use rebase_path you can drop this.

**Review by dnfield — APPROVED** [review]
LGTM with nits

File: testing/dart/canvas_test.dart (hunk 5 above)
Comment (matanlurey) [reply testing/dart/canvas_test.dart]: OH so that's what that means! Great, done.

**Comment by matanlurey** [comment]
@jonahwilliams I'm going to mark as auto-submit, but I'm happy to take additional comments post-submit.

**Comment by jonahwilliams** [comment]
LGTM!

**Comment by matanlurey** [comment]
The fonts are still showing up blank on Skia-gold, let me take a look again.

File: testing/dart/canvas_test.dart [hunk 6]
```diff
@@ -530,7 +540,7 @@ void main() async {
     // Skia renders a tofu if the font does not have a glyph for a character.
     // However, Flutter opts-in to a Skia feature to render tabs as a single space.
     // See: https://github.com/flutter/flutter/issues/79153
-    final File file = File(path.join('flutter', 'testing', 'resources', 'RobotoSlab-VariableFont_wght.ttf'));
+    final File file = File(path.join(_flutterBuildPath, 'flutter', 'third_party', 'txt', 'assets', 'Roboto-Regular.ttf'));
```
Comment (jonahwilliams) [inline testing/dart/canvas_test.dart:543]: I would expect this to be in a folder called fixtures.

Comment (jonahwilliams) [reply testing/dart/canvas_test.dart:543]: I think you want "./gen/flutter/impeller/fixtures/assets/Roboto-Regular.ttf"

File: testing/dart/canvas_test.dart (hunk 3 above)
Comment (matanlurey) [reply testing/dart/canvas_test.dart]: Done.

File: testing/dart/BUILD.gn (hunk 4 above)
Comment (matanlurey) [reply testing/dart/BUILD.gn]: Done.

File: testing/dart/canvas_test.dart (hunk 5 above)
Comment (matanlurey) [reply testing/dart/canvas_test.dart]: Done.

File: testing/dart/canvas_test.dart (hunk 6 above)
Comment (matanlurey) [reply testing/dart/canvas_test.dart:543]: But my `deps` is:
```gn
deps = [ "//flutter/third_party/txt:txt_fixtures" ]
```
What does that have to do with the impeller folder?

Comment (matanlurey) [reply testing/dart/canvas_test.dart:543]:
```bash
$ flutter % ls ../out/host_debug_unopt_arm64/gen/flutter/third_party/txt/assets                   
Roboto-Regular.ttf
```
Hmm....

Comment (jonahwilliams) [reply testing/dart/canvas_test.dart:543]: I'm not sure if this is the right target specifically, but any test fixtures should be built to a target in the out directory that you can look up

Comment (matanlurey) [reply testing/dart/canvas_test.dart:543]: Also if the file didn't exist, we could get a FileNotFoundError AFAICT

Comment (jonahwilliams) [reply testing/dart/canvas_test.dart:543]: ahh yeah. Hmmm

Comment (jonahwilliams) [reply testing/dart/canvas_test.dart:543]: I mean, the decorations _are_ showing up right 😆

Comment (matanlurey) [reply testing/dart/canvas_test.dart:543]: Yeah, though that's also true if the font doesn't load (hey, I don't make the rules)

I'll take another look in a bit and add some FML_LOG statements and see if there is something obvious going wrong. Probably (?) using a font in a test that doesn't exist should be an error or at the very least an angry log statement.

**Comment by matanlurey** [comment]
> The fonts are still showing up blank on Skia-gold, let me take a look again.

@jonahwilliams @dnfield Turns out this is because the default font color is white on white on Skia gold, and had nothing to do with fonts. I'm now outputting blue and it looks good, PTAL.

**Comment by jonahwilliams** [comment]
lol wonderful. Well now we know

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Pull request merged**